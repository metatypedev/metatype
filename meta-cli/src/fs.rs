// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use anyhow::{bail, Result};
use pathdiff::diff_paths;
use std::path::{Path, PathBuf};

#[cfg(target_os = "windows")]
pub fn is_hidden(path: impl AsRef<Path>) -> bool {
    use std::os::windows::fs::MetadataExt;

    // https://learn.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants
    path.as_ref()
        .metadata()
        .map(|m| m.file_attributes() & 0x02 == 0x02)
        .unwrap_or(false)
}

#[cfg(not(target_os = "windows"))]
pub fn is_hidden(path: impl AsRef<Path>) -> bool {
    path.as_ref()
        .file_name()
        .and_then(|n| n.to_str())
        .map_or(false, |n| n.starts_with('.'))
}

pub fn find_in_parents<P: AsRef<Path>>(start_dir: P, files: &[&str]) -> Result<Option<PathBuf>> {
    let mut current_dir = start_dir.as_ref().to_path_buf();
    if !current_dir.is_dir() {
        current_dir.pop();
    }
    loop {
        for file in files.iter() {
            let path = current_dir.join(file);
            if path.try_exists()? {
                return Ok(Some(path));
            }
        }
        if !current_dir.pop() {
            return Ok(None);
        }
    }
}

pub fn clean_path<P: AsRef<Path>, B: AsRef<Path>>(base_dir: B, path: P) -> Result<String> {
    let base_dir = base_dir.as_ref();
    let path = path.as_ref();
    let path = match diff_paths(path, base_dir) {
        Some(rel) => rel,
        None => {
            if base_dir.join(path).try_exists()? {
                path.to_path_buf()
            } else {
                bail!(
                    "Path {} is not in the base directory {} or does not exist",
                    path.display(),
                    base_dir.display()
                )
            }
        }
    };
    Ok(path.display().to_string())
}
