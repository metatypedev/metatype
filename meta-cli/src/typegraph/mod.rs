// Copyright Metatype OÃœ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

pub mod dependency_graph;
pub mod loader;
pub mod postprocess;
pub mod push;
pub mod utils;
