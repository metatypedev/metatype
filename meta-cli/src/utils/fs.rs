// Copyright Metatype OÃœ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

use std::path::Path;

#[cfg(target_os = "windows")]
pub fn is_hidden(path: impl AsRef<Path>) -> bool {
    use std::os::windows::fs::MetadataExt;

    // https://learn.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants
    path.as_ref()
        .metadata()
        .map(|m| m.file_attributes() & 0x02 == 0x02)
        .unwrap_or(false)
}

#[cfg(not(target_os = "windows"))]
pub fn is_hidden(path: impl AsRef<Path>) -> bool {
    path.as_ref()
        .file_name()
        .and_then(|n| n.to_str())
        .map_or(false, |n| n.starts_with('.'))
}
