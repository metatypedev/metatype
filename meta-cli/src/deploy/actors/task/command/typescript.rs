// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::interlude::*;
use tokio::process::Command;

enum TsRuntime {
    Deno,
    Node,
    Bun,
}

// TODO no `x tsx` for .js or .mjs files
pub(super) async fn get_raw_command(path: impl AsRef<Path>) -> Result<Command> {
    let path = path.as_ref();
    match detect_runtime(path).await? {
        TsRuntime::Deno => {
            log::debug!("loading typegraph using deno");
            let mut command = Command::new("deno");
            command
                .arg("run")
                // .arg("--unstable")
                .arg("--allow-all")
                .arg("--check")
                .arg(path.to_str().unwrap());
            Ok(command)
        }
        TsRuntime::Node => {
            log::debug!(
                "loading typegraph using npm x tsx, make sure npm packages have been installed"
            );
            let mut command = Command::new("npm");
            command
                .arg("x")
                .arg("--yes")
                .arg("tsx")
                .arg(path.to_str().unwrap());
            Ok(command)
        }
        TsRuntime::Bun => {
            log::debug!(
                "loading typegraph using bun x tsx, make sure npm packages have been installed"
            );
            let mut command = Command::new("bun");
            command
                .arg("x")
                .arg("tsx")
                .arg(path.to_str().unwrap())
                .current_dir(path.parent().unwrap());
            Ok(command)
        }
    }
}

// TODO cache?
async fn detect_runtime(tg_path: &Path) -> Result<TsRuntime> {
    use utils::*;
    use TsRuntime::*;

    let mut maybe_parent_dir = tg_path.parent();
    // try to detect runtime in use by checking for package.json/deno.json
    // files first
    loop {
        let Some(parent_dir) = maybe_parent_dir else {
            break;
        };
        log::trace!("testing for ts project manifest in {parent_dir:?}");
        if has_deno_json(parent_dir).await {
            log::trace!("deno.json hit in {parent_dir:?}");
            return Ok(Deno);
        }
        if has_package_json(parent_dir).await {
            log::trace!("package.json hit in {parent_dir:?}");
            if test_node_exec().await? {
                return Ok(Node);
            }
            if test_bun_exec().await? {
                return Ok(Bun);
            }
        }
        maybe_parent_dir = parent_dir.parent();
    }

    // if no package manifest found, just use the first runtime found in the
    // following order
    if test_deno_exec().await? {
        return Ok(Deno);
    }
    if test_node_exec().await? {
        return Ok(Node);
    }
    if test_bun_exec().await? {
        return Ok(Bun);
    }
    Err(ferr!("unable to find deno, node or bun runtimes"))
}

mod utils {
    use super::*;
    use cached::proc_macro::cached;

    #[cached(result = true)]
    pub async fn test_deno_exec() -> Result<bool> {
        Ok(Command::new("deno")
            .arg("--version")
            .output()
            .await
            .map(|out| out.status.success())?)
    }

    #[cached(result = true)]
    pub async fn test_node_exec() -> Result<bool> {
        Ok(Command::new("node")
            .arg("-v")
            .output()
            .await
            .map(|out| out.status.success())?)
    }

    #[cached(result = true)]
    pub async fn test_bun_exec() -> Result<bool> {
        Ok(Command::new("bun")
            .arg("-v")
            .output()
            .await
            .map(|out| out.status.success())?)
    }

    #[inline]
    pub async fn has_deno_json(dir: &Path) -> bool {
        use tokio::fs::try_exists;
        matches!(try_exists(dir.join("deno.json")).await, Ok(true))
            || matches!(try_exists(dir.join("deno.jsonc")).await, Ok(true))
    }

    #[inline]
    pub async fn has_package_json(dir: &Path) -> bool {
        use tokio::fs::try_exists;
        matches!(try_exists(dir.join("package.json")).await, Ok(true))
    }
}
