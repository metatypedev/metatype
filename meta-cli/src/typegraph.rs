// Copyright Metatype OÃœ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

use anyhow::{bail, Context, Result};
use colored::Colorize;
use common::typegraph::{FunctionMatData, Materializer, ModuleMatData, Typegraph};
use pathdiff::diff_paths;
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use walkdir::{DirEntry, WalkDir};

use crate::config::Config;
use crate::utils::ensure_venv;
use typescript::parser::{transform_module, transform_script};

pub type LoaderResult = HashMap<String, Result<Vec<Typegraph>>>;

#[derive(Clone)]
pub struct TypegraphLoader<'a> {
    skip_deno_modules: bool,
    ignore_unknown_file_types: bool,
    config: &'a Config,
}

impl<'a> TypegraphLoader<'a> {
    // pub fn new() -> Self {
    //     Self {
    //         skip_deno_modules: false,
    //         ignore_unknown_file_types: false,
    //         config: None,
    //     }
    // }

    pub fn with_config(config: &'a Config) -> Self {
        Self {
            skip_deno_modules: false,
            ignore_unknown_file_types: false,
            config,
        }
    }

    pub fn skip_deno_modules(mut self) -> Self {
        self.skip_deno_modules = true;
        self
    }

    // Loading a file shall fail if the file type is unsupported.
    // This is used when walking through directories, where unknown file types
    // should be skipped.
    pub fn ignore_unknown_file_types(mut self) -> Self {
        self.ignore_unknown_file_types = true;
        self
    }

    pub fn load_file<P: AsRef<Path>>(self, path: P) -> Result<Option<Vec<Typegraph>>> {
        let path = path.as_ref().canonicalize()?;
        let ext = path.extension().and_then(|ext| ext.to_str());

        let tgs = match ext {
            Some(ext) if ext == "py" => self
                .load_python_module(&path)
                .with_context(|| format!("Loading python module {:?}", path))?,
            _ => {
                if self.ignore_unknown_file_types {
                    return Ok(None);
                } else {
                    let ext = ext
                        .map(|ext| format!(".{ext}"))
                        .unwrap_or_else(|| "".to_owned());
                    bail!("Unsupported typegraph definition module with extension \"{ext}\": current version only support Python modules.");
                }
            }
        };

        let tgs: Vec<Typegraph> = serde_json::from_str(&tgs)?;
        Ok(Some(
            tgs.into_iter()
                .map(postprocess)
                .collect::<Result<Vec<_>>>()?,
        ))
    }

    pub fn load_files(self, files: &[PathBuf]) -> LoaderResult {
        files
            .iter()
            .filter_map(|file| {
                let path = match file.canonicalize() {
                    Ok(path) => path,
                    Err(e) => {
                        return Some((
                            file.to_str().unwrap().to_owned(),
                            Err(e).with_context(|| {
                                format!("could not canonicalize path: {:?}", file)
                            }),
                        ))
                    }
                };
                match self.clone().load_file(path) {
                    Ok(None) => None, // unreachable case
                    Ok(Some(tgs)) => Some((file.to_str().unwrap().to_owned(), Ok(tgs))),
                    Err(e) => Some((file.to_str().unwrap().to_owned(), Err(e))),
                }
            })
            .collect()
    }

    pub fn load_folder<P: AsRef<Path>>(self, dir: P) -> Result<LoaderResult> {
        let dir = dir.as_ref().canonicalize()?;
        let metadata =
            fs::metadata(&dir).with_context(|| format!("Reading the metadata of {:?}", dir))?;
        if !metadata.is_dir() {
            bail!("Expected a directory");
        }

        let loader = self.ignore_unknown_file_types();

        let py_loader = loader.config.loader("python").unwrap(); // cannot be none
        let include_set = py_loader.get_include_set()?;
        let exclude_set = py_loader.get_exclude_set()?;

        Ok(WalkDir::new(dir)
            .into_iter()
            .filter_entry(|e| !is_hidden(e))
            .filter_map(|e| {
                // filter files
                e.ok().map(|e| e.path().to_owned()).filter(|path| {
                    fs::metadata(path)
                        .ok()
                        .map(|m| m.is_file())
                        .unwrap_or(false)
                })
            })
            .filter_map(|path| {
                // inclusion/exclusion
                let relative = diff_paths(&path, &loader.config.base_dir).unwrap();
                let included = include_set.is_empty() || include_set.is_match(&relative);
                let excluded = !exclude_set.is_empty() && exclude_set.is_match(&relative);
                if included && !excluded {
                    println!(
                        "{}",
                        format!("Found typegraph definition module at {path:?}").dimmed()
                    );
                    Some(path)
                } else {
                    None
                }
            })
            .filter_map(|file| match loader.clone().load_file(&file) {
                Ok(None) => None,
                Ok(Some(tgs)) => Some((file.to_str().unwrap().to_owned(), Ok(tgs))),
                Err(e) => Some((file.to_str().unwrap().to_owned(), Err(e))),
            })
            .collect())
    }

    // Language-specific steps.
    // Returning typegraphs in raw (before post-processing) JSON.

    pub fn load_python_module<P: AsRef<Path>>(self, path: P) -> Result<String> {
        ensure_venv(&self.config.base_dir)?;

        // Search in PATH does not work on Windows
        // See: https://doc.rust-lang.org/std/process/struct.Command.html#method.new
        #[cfg(target_os = "windows")]
        let program_name = Path::new(&env::var("VIRTUAL_ENV")?).join("Scripts/py-tg.cmd");
        #[cfg(not(target_os = "windows"))]
        let program_name = Path::new("py-tg").to_path_buf();

        let p = Command::new(program_name.clone())
            .arg(path.as_ref().to_str().unwrap())
            // .args(args)
            .current_dir(&self.config.base_dir)
            .envs(env::vars())
            .env("PYTHONUNBUFFERED", "1")
            .env("PYTHONDONTWRITEBYTECODE", "1")
            .env(
                "DONT_READ_EXTERNAL_TS_FILES",
                if self.skip_deno_modules { "1" } else { "" },
            )
            .output()
            .with_context(|| {
                format!(
                    "Running the command '{:?} {:?}'",
                    program_name,
                    path.as_ref()
                )
            })?;

        if p.status.success() {
            Ok(String::from_utf8(p.stdout)?)
        } else {
            let stderr = String::from_utf8(p.stderr)?;
            bail!("Python error:\n{}", stderr.red())
        }
    }
}

type MaterializerPostprocessor = fn(Materializer, &Typegraph) -> Result<Materializer>;

static MATERIALIZER_POSTPROCESSORS: &[MaterializerPostprocessor] =
    &[postprocess_function_mat, postprocess_module_mat];

/// Perform some postprocessing on the typegraph we got from Python
fn postprocess(mut typegraph: Typegraph) -> Result<Typegraph> {
    let materializers = std::mem::take(&mut typegraph.materializers);
    let postprocessed_materializers: Vec<Materializer> = materializers
        .into_iter()
        .map(|mat| -> Result<Materializer> {
            let mut current_value = mat;
            for postprocess in MATERIALIZER_POSTPROCESSORS {
                current_value = postprocess(current_value, &typegraph)?;
            }
            Ok(current_value)
        })
        .collect::<Result<Vec<Materializer>>>()?;
    typegraph.materializers = postprocessed_materializers;
    Ok(typegraph)
}

fn postprocess_function_mat(mut mat: Materializer, typegraph: &Typegraph) -> Result<Materializer> {
    if &mat.name == "function" && typegraph.runtimes[mat.runtime as usize].name == "deno" {
        let mut mat_data: FunctionMatData = utils::object_from_map(std::mem::take(&mut mat.data))?;
        // TODO check variable `_my_lambda` exists and is a function expression/lambda
        mat_data.script = transform_script(mat_data.script)?;
        mat.data = utils::map_from_object(mat_data)?;
    }
    Ok(mat)
}

fn postprocess_module_mat(mut mat: Materializer, typegraph: &Typegraph) -> Result<Materializer> {
    if mat.name == "module" && typegraph.runtimes[mat.runtime as usize].name == "deno" {
        let mut mat_data: ModuleMatData = utils::object_from_map(std::mem::take(&mut mat.data))?;
        if !mat_data.code.starts_with("file:") {
            // TODO check imported functions exist
            mat_data.code = transform_module(mat_data.code)?;
        }
        mat.data = utils::map_from_object(mat_data)?;
    }
    Ok(mat)
}

mod utils {
    use anyhow::{bail, Result};
    use indexmap::IndexMap;
    use serde::{de::DeserializeOwned, ser::Serialize};
    use serde_json::{from_value, to_value, Value};

    pub fn object_from_map<T: DeserializeOwned>(map: IndexMap<String, Value>) -> Result<T> {
        let map = Value::Object(map.into_iter().collect());
        Ok(from_value(map)?)
    }

    pub fn map_from_object<T: Serialize>(obj: T) -> Result<IndexMap<String, Value>> {
        let val = to_value(obj)?;
        if let Value::Object(map) = val {
            Ok(map.into_iter().collect())
        } else {
            bail!("value is not an object");
        }
    }
}

fn is_hidden(entry: &DirEntry) -> bool {
    entry
        .file_name()
        .to_str()
        .map(|s| s.starts_with('.'))
        .unwrap_or(false)
}
