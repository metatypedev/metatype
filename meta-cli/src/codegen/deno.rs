// Copyright Metatype OÜ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

use anyhow::{bail, Context, Result};
use colored::Colorize;
use common::typegraph::{TypeNode, Typegraph};
use log::{info, trace};
use pathdiff::diff_paths;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::cell::RefCell;
use std::collections::{HashMap, HashSet};
use std::fmt::Write as _;
use std::fs::{self, File};
use std::io::Write;
use std::path::{Path, PathBuf};
use typescript as ts;

#[derive(Serialize, Deserialize, Debug)]
struct ImportFuncMatData {
    name: String,
    #[serde(rename = "mod")]
    module: u32,
}

#[derive(Serialize, Deserialize, Debug)]
struct ModuleMatData {
    code: String,
}

// TODO implement as a post-processor
/// Generate codes for missing deno function referenced from the typegraph.
/// Returns false if no function is to be generated.
/// Parameter `meta_codegen` is true if the codegen in run from the `meta codegen` subcommand.
pub fn codegen<P>(tg: &Typegraph, base_dir: P) -> Result<bool>
where
    P: AsRef<Path>,
{
    let cg = Codegen::new(tg, base_dir);
    let codes = cg.codegen()?;
    if codes.is_empty() {
        return Ok(false);
    }

    let current_dir = std::env::current_dir().unwrap();
    for code in codes.into_iter() {
        let parent_folder = Path::new(&code.path).parent().unwrap();
        fs::create_dir_all(parent_folder).unwrap();

        let mut file = File::options()
            .create(true)
            .append(true)
            .open(&code.path)
            .context(format!("could not open output file: {:?}", code.path))?;
        write!(file, "\n{code}\n", code = code.code)?;
        let rel_path = diff_paths(&code.path, &current_dir).unwrap();
        info!(
            "{} Successfully added new function(s) in {:?}.",
            "✓".green(),
            rel_path
        );
    }
    Ok(true)
}

struct ModuleCode {
    path: PathBuf,
    code: String,
}

#[derive(Debug)]
struct ModuleInfo {
    path: PathBuf,
    exports: RefCell<Option<HashSet<String>>>, // names of exported functions
}

#[derive(Debug)]
struct GenItem {
    input: u32,
    output: u32,
    path: String,
    name: String,
}

pub struct Codegen<'a> {
    pub path: PathBuf,
    pub base_dir: PathBuf,
    tg: &'a Typegraph,
    ts_modules: HashMap<String, ModuleInfo>,
}

impl<'a> Codegen<'a> {
    fn new<P>(tg: &'a Typegraph, path: P) -> Self
    where
        P: AsRef<Path>,
    {
        let path = path.as_ref().to_path_buf();
        let base_dir = {
            let mut dir = path.clone();
            if !dir.pop() {
                panic!("invalid file path");
            }
            dir
        };

        let ts_modules = {
            let mut modules = HashMap::new();
            for mat in tg.materializers.iter() {
                let runtime = &tg.runtimes[mat.runtime as usize];
                if &runtime.name != "deno" && &runtime.name != "worker" {
                    continue;
                }
                if mat.name != "module" {
                    continue;
                }
                let code: String =
                    serde_json::from_value(mat.data.get("code").unwrap().clone()).unwrap();
                if let Some(relpath) = code.strip_prefix("file:") {
                    let path = {
                        let mut path = base_dir.clone();
                        // TODO is this necessary?? py-tg yields absolute path!!
                        path.push(relpath);
                        path
                    };
                    modules.insert(
                        relpath.to_string(),
                        ModuleInfo {
                            path,
                            exports: RefCell::new(None), // lazy
                        },
                    );
                }
            }
            modules
        };
        Codegen {
            path,
            base_dir,
            tg,
            ts_modules,
        }
    }

    fn codegen(mut self) -> Result<Vec<ModuleCode>> {
        let mut gen_list = vec![];

        for tpe in self.tg.types.iter() {
            if let TypeNode::Function {
                materializer,
                input,
                output,
                base,
                ..
            } = tpe
            {
                let mat = self.tg.materializers[*materializer as usize].clone();
                let runtime = &self.tg.runtimes[base.runtime as usize];
                if runtime.name != "deno" && runtime.name != "worker" {
                    continue;
                }
                if mat.name == "import_function" {
                    let mat_data: ImportFuncMatData =
                        serde_json::from_value(serde_json::to_value(mat.data)?)
                            .expect("invalid materializer data for function materializer");
                    let module_mat = &self.tg.materializers[mat_data.module as usize];
                    let path: String =
                        serde_json::from_value(module_mat.data.get("code").unwrap().clone())
                            .unwrap();
                    if let Some(path) = path.strip_prefix("file:") {
                        if self.ts_modules.contains_key(path)
                            && self.check_func(path, &mat_data.name)
                        {
                            trace!("Entry[{path:?}]: {:?}", self.ts_modules.get(path).unwrap());
                            gen_list.push(GenItem {
                                input: *input,
                                output: *output,
                                path: self
                                    .ts_modules
                                    .get(path)
                                    .unwrap()
                                    .path
                                    .to_str()
                                    .unwrap()
                                    .to_string(),
                                name: mat_data.name.clone(),
                            });
                        }
                    }
                }
            }
        }

        // group by modules
        let gen_list = gen_list;
        self.generate(gen_list)
            .into_iter()
            .map(|(name, code)| -> Result<ModuleCode> {
                let path = self.ts_modules.remove(&name).unwrap().path;
                trace!("Code path: {path:?}");
                let code = ts::format_text(&path, &code)
                    .context(format!("could not format code: {code:#?}"))?;
                Ok(ModuleCode { path, code })
            })
            .collect::<Result<Vec<_>>>()
    }

    fn generate(&self, gen_list: Vec<GenItem>) -> HashMap<String, String> {
        let mut map: HashMap<String, Vec<Result<String>>> = HashMap::default();
        let current_dir = std::env::current_dir().unwrap();
        for GenItem {
            input,
            output,
            path,
            name,
        } in gen_list.into_iter()
        {
            trace!("Codegen::generate: path={path:?}, current-dir={current_dir:?}");
            // let rel_path = diff_paths(&path, &current_dir).unwrap();
            info!("Generating missing function {} for {:?}", name.blue(), path);
            let functions = map.entry(path).or_default();
            functions.push(self.gen_func(input, output, &name));
        }

        let map = map
            .into_iter()
            .map(|(k, v)| (k, v.into_iter().collect::<Result<Vec<String>>>()))
            .collect::<HashMap<_, _>>();

        let mut cgs: HashMap<String, String> = HashMap::default();
        for (k, v) in map.into_iter() {
            cgs.insert(k, v.expect("could not generate code").join("\n\n"));
        }

        cgs
    }

    fn exported_functions<P>(mod_path: P) -> HashSet<String>
    where
        P: AsRef<Path>,
    {
        if !mod_path.as_ref().exists() {
            return HashSet::new();
        }
        let module = ts::parser::parse_module(mod_path.as_ref())
            .unwrap_or_else(|_| panic!("could not load module: {:?}", mod_path.as_ref()));

        ts::parser::get_exported_functions(&module.body)
    }

    /// Returns `true` if the function named `name` should be generated in the specified file.
    fn check_func(&self, path: &str, name: &str) -> bool {
        if let Some(module) = self.ts_modules.get(path) {
            if module.exports.borrow().is_some() {
                let exports = module.exports.borrow();
                let exports = exports.as_ref().unwrap();
                !exports.contains(name)
            } else {
                let exports = Self::exported_functions(&module.path);
                let ret = !exports.contains(name);
                *module.exports.borrow_mut() = Some(exports);
                ret
            }
        } else {
            false
        }
    }

    fn gen_interface(&self, name: &str, idx: u32) -> Result<String> {
        let tpe = &self.tg.types[idx as usize];
        Ok(format!("interface {name} {}\n", self.gen_obj_type(tpe)?))
    }

    fn gen_union_type_definition(&self, any_of: &[u32]) -> Result<String> {
        let mut variant_definitions = Vec::new();
        for &variant_type_index in any_of {
            let variant_type_definition = self
                .get_typespec(variant_type_index)
                .expect("type definition generation for variant type should be supported");
            variant_definitions.push(variant_type_definition);
        }
        // field `anyOf` in JSON Schema can be represented in TypeScript
        // as a union of all the variant types
        let intersection_type = variant_definitions.join(" | ");
        Ok(intersection_type)
    }

    fn gen_obj_type(&self, tpe: &TypeNode) -> Result<String> {
        let fields = tpe.get_struct_fields()?;
        let fields = fields
            .iter()
            .map(|(k, v)| (k.clone(), self.get_typespec(*v).unwrap()))
            .collect::<HashMap<_, _>>();
        #[cfg(test)]
        let fields = fields
            .into_iter()
            .collect::<std::collections::BTreeMap<_, _>>();
        let mut typedef = "{\n".to_string();
        for (k, v) in fields.iter() {
            writeln!(typedef, "  {k}: {v};")?;
        }
        typedef.push('}');
        Ok(typedef)
    }

    fn destructure_object(&self, idx: u32) -> Result<String> {
        let tpe = &self.tg.types[idx as usize];
        let fields = tpe.get_struct_fields()?;
        #[cfg(test)]
        let fields = fields
            .into_iter()
            .collect::<std::collections::BTreeMap<_, _>>();
        Ok(format!(
            "{{ {} }}",
            fields.keys().cloned().collect::<Vec<_>>().join(", "),
        ))
    }

    fn gen_default_value(&self, idx: u32) -> Result<String> {
        let tpe = &self.tg.types[idx as usize];
        match tpe {
            TypeNode::Optional { .. } => Ok("null".to_owned()),
            TypeNode::Array { .. } => Ok("[]".to_owned()),
            TypeNode::Boolean { .. } => Ok("false".to_owned()),
            TypeNode::Number { .. } | TypeNode::Integer { .. } => Ok("0".to_owned()),
            TypeNode::String { .. } => Ok("\"\"".to_owned()),
            TypeNode::Object { properties, .. } => {
                let props = properties.clone();
                #[cfg(test)]
                let props = props
                    .into_iter()
                    .collect::<std::collections::BTreeMap<_, _>>();
                let body = props
                    .iter()
                    .map(|(k, v)| -> Result<String> {
                        Ok(format!("{k}: {}", self.gen_default_value(*v)?))
                    })
                    .collect::<Result<Vec<_>>>()?
                    .join(", ");
                Ok(format!("{{ {body} }}"))
            }
            TypeNode::Union { any_of, .. } => {
                // a type cannot be all the variants, as they might be a
                // disjoint union, therefore by returning the default value of
                // one variant would be enough
                let variant_type_index = any_of
                    .clone()
                    .pop()
                    .expect("the union type should have at least one variant of type nodes");
                self.gen_default_value(variant_type_index)
            }
            TypeNode::Either { one_of, .. } => {
                // a type cannot be all the variants, as they are a
                // disjoint union, therefore by returning the default value of
                // one variant would be enough
                let variant_type_index = one_of
                    .clone()
                    .pop()
                    .expect("the either type should have at least one variant of type nodes");
                self.gen_default_value(variant_type_index)
            }
            _ => bail!("unsupported type to generate default value: {tpe:#?}"),
        }
    }

    fn gen_func(&self, input: u32, output: u32, name: &str) -> Result<String> {
        let type_name = {
            let mut name = name.to_string();
            let (lead, _) = name.split_at_mut(1);
            lead.make_ascii_uppercase();
            name
        };

        // input type
        let inp_type_name = format!("{type_name}Input");

        // let inp_type = &self.tg.types[fn_data.input as usize];
        let inp_typedef = self
            .gen_interface(&inp_type_name, input)
            .context("failed to generate input type")?;

        let output_type_node = &self.tg.types[output as usize];

        // variable helper in case out_typespec is an interface
        let mut output_type_definition: Option<String> = None;

        // for type definitions that are too long use interfaces/types
        // to avoid cluttering the function definition
        let out_typespec = {
            const LENGTH_LIMIT: usize = 35;
            let output_type_name = format!("{type_name}Output");

            match output_type_node {
                TypeNode::Union { .. } => {
                    let type_definition = self
                        .get_typespec(output)
                        .context("failed to generate type definition")?;

                    if type_definition.len() > LENGTH_LIMIT {
                        output_type_definition =
                            Some(format!("type {output_type_name} = {type_definition}\n"));
                        output_type_name
                    } else {
                        type_definition
                    }
                }
                _ => self
                    .get_typespec(output)
                    .context("failed to generate output type")?,
            }
        };

        let code = format!(
            "{}\n{}\nexport {}function {}({}: {}, {{ context }}: {{ context: Record<string, unknown> }}): {} {{\n  return {};\n}}",
            inp_typedef,
			output_type_definition.unwrap_or_default(),
            if name == "default" { "default " } else {""},
            if name == "default" { "" } else { name },
            self.destructure_object(input)?,
            inp_type_name,
            out_typespec,
            self.gen_default_value(output)?,
        );

        // TODO: format code
        Ok(code)
    }

    fn get_typespec(&self, idx: u32) -> Result<String> {
        let tpe = &self.tg.types[idx as usize];

        match tpe {
            TypeNode::Optional { item, .. } => Ok(format!("null | {}", self.get_typespec(*item)?)),
            TypeNode::Array { items, .. } => Ok(format!("Array<{}>", self.get_typespec(*items)?)),
            TypeNode::Boolean { .. } => Ok("boolean".to_owned()),
            TypeNode::Number { .. } | TypeNode::Integer { .. } => Ok("number".to_owned()),
            TypeNode::String { base, .. } => {
                if let Some(variants) = &base.enumeration {
                    let variants: Vec<String> = variants
                        .iter()
                        .map(|variant| {
                            let variant = variant
                                .as_str()
                                .expect("each variant of a string enum should be a string");

                            format!("\"{variant}\"")
                        })
                        .collect();

                    let enum_definition = variants.join(" | ");
                    Ok(enum_definition)
                } else {
                    Ok("string".to_owned())
                }
            }
            TypeNode::Object { .. } => self.gen_obj_type(tpe),
            TypeNode::Union { any_of, .. } => self.gen_union_type_definition(any_of),
            TypeNode::Either { one_of, .. } => self.gen_union_type_definition(one_of),
            _ => bail!("unsupported type to generate type specification: {tpe:#?}"),
        }
    }
}

/**
 * utils
 **/

trait IntoJson {
    fn into_json(self) -> Value;
}

impl IntoJson for HashMap<String, Value> {
    fn into_json(self) -> Value {
        let mut map = serde_json::Map::with_capacity(self.len());
        for (k, v) in self {
            map.insert(k, v);
        }
        Value::Object(map)
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use normpath::PathExt;

    use super::*;
    use crate::config::Config;
    use crate::tests::utils::ensure_venv;
    use crate::typegraph::loader::{Loader, LoaderResult};

    #[tokio::test(flavor = "multi_thread")]
    async fn codegen() -> Result<()> {
        crate::logger::init();
        ensure_venv()?;
        let test_folder = Path::new("./src/tests/typegraphs").normalize()?;
        std::env::set_current_dir(&test_folder)?;
        trace!("Test folder: {test_folder:?}");
        let tests = fs::read_dir(&test_folder).unwrap();
        let config = Config::default_in(".");
        let config = Arc::new(config);

        for typegraph_test in tests {
            let typegraph_test = typegraph_test.unwrap().path();
            let typegraph_test = diff_paths(&typegraph_test, &test_folder).unwrap();
            trace!("test: {typegraph_test:?}");
            let loader = Loader::new(Arc::clone(&config)).skip_deno_modules(true);

            match loader.load_file(&typegraph_test).await {
                LoaderResult::Loaded(tgs) => {
                    assert_eq!(tgs.len(), 1);
                    let tg = &tgs[0];

                    let module_codes = Codegen::new(tg, &typegraph_test).codegen()?;
                    assert_eq!(module_codes.len(), 1);

                    let test_name = typegraph_test.to_string_lossy().to_string();
                    trace!("test-name={test_name:?}");
                    insta::assert_snapshot!(test_name, &module_codes[0].code);
                }
                LoaderResult::Error(e) => {
                    bail!(
                        "Error while loading typegraph from {typegraph_test:?}: {e}",
                        e = e.to_string()
                    );
                }
                LoaderResult::Rewritten(_) => {
                    bail!("Unexpected: typegraph definition module has been rewritten");
                }
            }
        }

        Ok(())
    }
}
