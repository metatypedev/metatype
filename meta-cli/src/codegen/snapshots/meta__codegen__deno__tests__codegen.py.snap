---
source: meta-cli/src/codegen/deno.rs
expression: "&module_codes[0].code"
---
interface DivInput {
  dividend: number;
  divisor: number;
}

export function div({ dividend, divisor }: DivInput,
  { context }: { context: Record<string, unknown> }
): {
  quotient: number;
  remainder: number;
} {
  return { quotient: 0, remainder: 0 };
}
